query/services/query.py

```diff
@@ -0,0 +1,240 @@
+            )
+
+        if (
+            len(query.sort) != 1
+            or query.sort[0][0] != "popularity_key"
+            or query.sort[0][1] != -1
+        ):
+            raise HTTPException(
+                status.HTTP_422_UNPROCESSABLE_ENTITY,
+                "Only descending popularity_key sort is currently supported",
+            )
+
+        country_tag = getattr(query.filter, "countries-tags")
+        country = (
+            await get_country(conn, country_tag)
+            if country_tag
+            else await get_country(conn, "en:world")
+        )
+        if country_tag:
+            delattr(query.filter, "countries-tags")
+
+        sql_fragments = []
+        params = [country["id"]]
+        if query.filter:
+            append_sql_fragments(
+                query.filter,
+                loaded_tags,
+                "product_id",
+                params,
+                sql_fragments,
+            )
+        limit_clause = ""
+        if query.limit:
+            params.append(query.limit)
+            limit_clause += f" LIMIT ${len(params)}"
+        if query.skip:
+            params.append(query.skip)
+            limit_clause += f" OFFSET ${len(params)}"
+
+        sql = f"""SELECT p.code FROM product p 
+            JOIN product_country pc ON pc.product_id = p.id AND pc.country_id = $1
+            WHERE p.id IN (SELECT p.product_id
+                FROM product_country p
+                WHERE p.country_id = $1 
+                AND {'' if obsolete else 'NOT '}p.obsolete
+                {''.join(sql_fragments)}
+                ORDER BY p.recent_scans DESC, p.total_scans DESC, p.product_id
+                {limit_clause})
+            ORDER BY pc.recent_scans DESC, pc.total_scans DESC, pc.product_id"""
+        results = await fetch_and_log(conn, sql, *params)
+        product_codes = [result["code"] for result in results]
+        logger.debug(f"Find: Codes: {repr(product_codes)}")
+
+        if "code" not in query.projection:
+            query.projection["code"] = 1
+
+        if len(query.projection.keys()) == 1:
+            # Only requesting code so we don't need to go to MongoDB. Could extend this for other fields we store in off-query
+            return results
+        else:
+            mongodb_filter = {"_id": {"$in": product_codes}}
+            mongodb_results = [None] * len(product_codes)
+            async with find_products(
+                mongodb_filter, query.projection, obsolete
+            ) as cursor:
+                async for result in cursor:
+                    code_index = product_codes.index(result["code"])
+                    mongodb_results[code_index] = result
+
+            # Eliminate any None's from the result. Note this should only happen if there is a mismatch between off-query and MongoDB
+            return [result for result in mongodb_results if result]
+
+
+def append_sql_fragments(
```

Comment:
This function does not yet have unit test, while it might be worth it.

Reply:
I'm worried that might be brittle as we would be doing text comparison on the SQL where what we want is the correct results. I have checked and we already have 100% overage with the existing tests